design/src/rfcs/rfc0027_improve_sdk_request_id_access.md

```diff
@@ -0,0 +1,244 @@
+request ID retrieval:
+
+```rust
+pub trait RequestId {
+    /// Returns the request ID if it's available.
+    fn request_id(&self) -> Option<&str>;
+}
+```
+
+This trait will be implemented for `SdkError` in `aws-http` where it is declared,
+complete with logic to pull the request ID header out of the raw HTTP responses
+(it will always return `None` for event stream `Message` responses; an additional
+trait may need to be added to `aws-smithy-http` to facilitate access to the headers).
+This logic will try different request ID header names in order of probability
+since AWS services have a couple of header name variations. `x-amzn-requestid` is
+the most common, with `x-amzn-request-id` being the second most common.
+
+`aws-http` will also implement `RequestId` for `aws_smithy_types::error::Error`,
+and the `request_id` method will be removed from `aws_smithy_types::error::Error`.
+Places that construct `Error` will place the request ID into its `extras` field,
+where the `RequestId` trait implementation can retrieve it.
+
+A codegen decorator will be added to `sdk-codegen` to implement `RequestId` for
+operation errors, and the existing `request_id` accessors will be removed from
+`CombinedErrorGenerator` in `codegen-core`.
+
+With these changes, customers can directly access request IDs from `SdkError` and
+operations errors by importing the `RequestId` trait. Additionally, the Smithy/SDK
+purity is improved since both places where request IDs are leaked to Smithy clients
+will be resolved.
+
+### Implement `RequestId` for outputs
+
+To make it possible to retrieve request IDs when using the fluent client, the new
+`RequestId` trait can be implemented for outputs.
+
+Some services (e.g., Transcribe Streaming) model the request ID header in their
+outputs, while other services (e.g., Directory Service) model a request ID
+field on errors. In some cases, services take `RequestId` as a modeled input
+(e.g., IoT Event Data). It follows that it is possible, but unlikely, that
+a service could have a field named `RequestId` that is not the same concept
+in the future.
+
+Thus, name collisions are going to be a concern for putting a request ID accessor
+on output. However, if it is implemented as a trait, then this concern is partially
+resolved. In the vast majority of cases, importing `RequestId` will provide the
+accessor without any confusion. In cases where it is already modeled and is the
+same concept, customers will likely just use it and not even realize they didn't
+import the trait. The only concern is future cases where it is modeled as a
+separate concept, and as long as customers don't import `RequestId` for something
+else in the same file, that confusion can be avoided.
+
+In order to implement `RequestId` for outputs, either the original response needs
+to be stored on the output, or the request ID needs to be extracted earlier and
+stored on the output. The latter will lead to a small amount of header lookup
+code duplication.
```

Comment:
although we can make an inlineable to do it—this is probably also where we can deal with any service-specific request ID behavior.